src/chainparams.cpp

```diff
@@ -474,14 +474,51 @@ class CRegTestParams : public CChainParams {
     /**
      * Allows modifying the Version Bits regtest parameters.
      */
-    void UpdateVersionBitsParameters(Consensus::DeploymentPos d, int64_t nStartTime, int64_t nTimeout)
+    void UpdateVersionBitsParameters(Consensus::DeploymentPos d, int64_t startheight, int64_t timeoutheight)
     {
-        consensus.vDeployments[d].nStartTime = nStartTime;
-        consensus.vDeployments[d].nTimeout = nTimeout;
+        consensus.vDeployments[d].startheight = startheight;
+        consensus.vDeployments[d].timeoutheight = timeoutheight;
     }
     void UpdateActivationParametersFromArgs(const ArgsManager& args);
 };
 
+bool CChainParams::CheckVBitsHeights(std::string& error, int64_t startheight, int64_t timeoutheight) const
+{
+    // Special always or never active cases
+    if ((startheight == Consensus::BIP9Deployment::NEVER_ACTIVE && timeoutheight == Consensus::BIP9Deployment::NEVER_ACTIVE)
+        || startheight == Consensus::BIP9Deployment::ALWAYS_ACTIVE) {
```

Comment:
Any reason to not enforce NO_TIMEOUT here?

Reply:
Not well-defined, not very friendly (it's  a max int value), and ignored anyway. (third time addressing this)

Reply:
In `GetStateFor`, we are checking that both `startheight` and `timeoutheight` are `NEVER_ACTIVE` in order to return the correct state for never active. It only checks `startheight` is `ALWAYS_ACTIVE` when determining always active. This behavior came from #19573 and I didn't change it. Perhaps @luke-jr or @ajtowns could explain why that is.

Reply:
It's simply a non-change from the original BIP9 code.

I don't see a reason to change it here.

Reply:
ALWAYS and NEVER are not specified in any BIP, they are special magic values in our code only (introduced in this pull request). Treating them in the same fashion will be less code, and less code complexity.

Reply:
NO_TIMEOUT is not introduced in this PR. Feel free to refactor in a followup... or achow can do it here if you care so strongly. I don't see that it matters either way.

Reply:
I think we can leave this for a followup.